### DISCUSSION TITLE
Lesson 11:  FileNotFoundError: [Errno 2] No such file or directory: './metadata/rinkeby/0-PUG.json'

### DISCUSSION THREAD

**Discussion by ChristianG77** [smartcontractkit/full-blockchain-solidity-course-py#241] [Q&A] [answered]
Hi,

On rinkeby, I've run deploy_and_create.py, create_collectible.py and waited for the tokenCounter = 2, then ran create_metadata.py and get the following error where it can't find the image.

PS C:\Users\chris\Dropbox\00_CODE\NFT-DEMO> brownie run scripts/advanced_collectible/create_metadata.py --network rinkeby
INFO: Could not find files for the given pattern(s).
Brownie v1.16.4 - Python development framework for Ethereum

NftDemoProject is the active project.

Running 'scripts\advanced_collectible\create_metadata.py::main'...
You have created 2 collectibles!
Creating Metadata file: ./metadata/rinkeby/0-PUG.json
  File "C:\Users\chris\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\run.py", line 49, in main
    return_value, frame = run(
  File "C:\Users\chris\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File ".\scripts\advanced_collectible\create_metadata.py", line 40, in main
    with open(metadata_file_name, "w") as file:
FileNotFoundError: [Errno 2] No such file or directory: './metadata/rinkeby/0-PUG.json'
PS C:\Users\chris\Dropbox\00_CODE\NFT-DEMO>

Below are the outputs from deploy_and _create.py, create_collectible.py and create_metadata.py.

PS C:\Users\chris\Dropbox\00_CODE\NFT-DEMO> brownie run scripts/advanced_collectible/deploy_and_create.py --network rinkeby 
INFO: Could not find files for the given pattern(s).
Brownie v1.16.4 - Python development framework for Ethereum

NftDemoProject is the active project.

Running 'scripts\advanced_collectible\deploy_and_create.py::main'...
Transaction sent: 0xcb51935b21fc8e955e44c49aef78d7b985ccd4f68becddbcc23c6c2252e174a3
  Gas price: 1.000000009 gwei   Gas limit: 2431924   Nonce: 32
  AdvancedCollectible deployed at: 0x846A06e5716517Db40F02398173C8a1D79AE186B

Waiting for https://api-rinkeby.etherscan.io/api to process contract...
Verification submitted successfully. Waiting for result...
Verification pending...
Verification complete. Result: Pass - Verified
Transaction sent: 0xeeebd12322f4d991533685abf7704d712eb7de74a7595f8532b2ed5019fced95
  Gas price: 1.000000009 gwei   Gas limit: 56992   Nonce: 33
  LinkToken.transfer confirmed   Block: 9505482   Gas used: 51811 (90.91%)

  LinkToken.transfer confirmed   Block: 9505482   Gas used: 51811 (90.91%)

Funded 0x846A06e5716517Db40F02398173C8a1D79AE186B
Transaction sent: 0xed15a6fbb8d539e4fbabf3e0bebfe350e37ab7cb66c3415c7c53d0cc84398a2e
  Gas price: 1.000000009 gwei   Gas limit: 190972   Nonce: 34
  AdvancedCollectible.createCollectible confirmed   Block: 9505483   Gas used: 173611 (90.91%)

  AdvancedCollectible.createCollectible confirmed   Block: 9505483   Gas used: 173611 (90.91%)

New token has been created!
INFO: Could not find files for the given pattern(s).
Brownie v1.16.4 - Python development framework for Ethereum

NftDemoProject is the active project.

Running 'scripts\advanced_collectible\create_collectible.py::main'...
Transaction sent: 0xe445ee6fc07c03c6b927ef991da03ef325ccf2f70e3dc52f0a424149940e87a9
  Gas price: 1.00000001 gwei   Gas limit: 38182   Nonce: 35
  LinkToken.transfer confirmed   Block: 9505513   Gas used: 34711 (90.91%)

  LinkToken.transfer confirmed   Block: 9505513   Gas used: 34711 (90.91%)

Funded 0x846A06e5716517Db40F02398173C8a1D79AE186B
Transaction sent: 0xcae6de95042e34db87b0746b8200c66958623649f771c0a332276178ed8d6805
  Gas price: 1.00000001 gwei   Gas limit: 153352   Nonce: 36

Collectible created!
PS C:\Users\chris\Dropbox\00_CODE\NFT-DEMO> brownie run scripts/advanced_collectible/create_metadata.py --network rinkeby
INFO: Could not find files for the given pattern(s).
Brownie v1.16.4 - Python development framework for Ethereum

NftDemoProject is the active project.

Running 'scripts\advanced_collectible\create_metadata.py::main'...
You have created 2 collectibles!
Creating Metadata file: ./metadata/rinkeby/0-PUG.json
  File "C:\Users\chris\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\_cli\run.py", line 49, in main
    return_value, frame = run(
  File "C:\Users\chris\.local\pipx\venvs\eth-brownie\lib\site-packages\brownie\project\scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File ".\scripts\advanced_collectible\create_metadata.py", line 40, in main
    with open(metadata_file_name, "w") as file:
FileNotFoundError: [Errno 2] No such file or directory: './metadata/rinkeby/0-PUG.json'

**Comment by cromewar** [comment]
Please post your create_metadata.py code here.

**Reply by ChristianG77** [reply]
from brownie import AdvancedCollectible, network
from scripts.helpful_scripts import get_breed
from metadata.sample_metadata import metadata_template
from pathlib import Path
import requests
import json
import os

breed_to_image_uri = {
    "PUG": "https://ipfs.io/ipfs/QmSsYRx3LpDAb1GZQm7zZ1AuHZjfbPkD6J7s9r41xu1mf8?filename=pug.png",
    "SHIBA_INU": "https://ipfs.io/ipfs/QmYx6GsYAKnNzZ9A6NvEKV9nf1VaDzJrqDR23Y8YSkebLU?filename=shiba-inu.png",
    "ST_BERNARD": "https://ipfs.io/ipfs/QmUPjADFGEKmfohdTaNcWhp7VGk26h5jXDA7v3VtTnTLcW?filename=st-bernard.png",
}


def main():
    advanced_collectible = AdvancedCollectible[-1]
    number_of_advanced_collectibles = advanced_collectible.tokenCounter()
    print(f"You have created {number_of_advanced_collectibles} collectibles!")
    for token_id in range(number_of_advanced_collectibles):
        breed = get_breed(advanced_collectible.tokenIdToBreed(token_id))
        metadata_file_name = (
            f"./metadata/{network.show_active()}/{token_id}-{breed}.json"
        )
        collectible_metadata = metadata_template
        if Path(metadata_file_name).exists():
            print(f"{metadata_file_name} already exists! Delete it to overwrite")
        else:
            print(f"Creating Metadata file: {metadata_file_name}")
            collectible_metadata["name"] = breed
            collectible_metadata["description"] = f"An adorable {breed} pup!"
            image_path = "./img/" + breed.lower().replace("_", "-") + ".png"

            image_uri = None
            if os.getenv("UPLOAD_IPFS") == "true":
                image_uri = upload_to_ipfs(image_path)
            image_uri = image_uri if image_uri else breed_to_image_uri[breed]

            collectible_metadata["image"] = image_uri
            with open(metadata_file_name, "w") as file:
                json.dump(collectible_metadata, file)
            if os.getenv("UPLOAD_IPFS") == "true":
                upload_to_ipfs(metadata_file_name)


def upload_to_ipfs(filepath):
    with Path(filepath).open("rb") as fp:
        image_binary = fp.read()
        ipfs_url = "http://127.0.0.1:5001"  # Need to get this from the terminal by inputting 'ipfs daemon'
        endpoint = "/api/v0/add"
        response = requests.post(ipfs_url + endpoint, files={"file": image_binary})
        ipfs_hash = response.json()["Hash"]
        # "./img/0-PUG.png" -> "0-PUG.png"
        filename = filepath.split("/")[-1:][0]
        image_uri = f"https://ipfs.io/ipfs/{ipfs_hash}?filename={filename}"
        # "https://ipfs.io/ipfs/Qmd9MCGtdVz2miNumBHDbvj8bigSgTwnr4SbyH6DNnpWdt?filename=0-PUG.json"
        print(image_uri)
        return image_uri

**Comment by cromewar** [comment, marked as answer]
it seems that everything is ok, however I wonder did you update the `.env` file and exported UPLOAD_IPFS, remember even you add it to .env sometimes you should export it manually in order to work.
Finally this is not important but i strongly suggest yo change this line: 
`image_uri = f"https://ipfs.io/ipfs/{ipfs_hash}?filename={filename}"`
for this:
`image_uri = "https://ipfs.io/ipfs/{}?filename={}".format(ipfs_hash, filename)`

**Reply by ChristianG77** [reply]
Thanks @cromewar for helping me out on this!  The problem was I was missing a folder under the 'metadata' hierarchy for rinkeby and then it worked!  Thanks again!

**Reply by cromewar** [reply]
Oh so you weren't pointing to the correct folder, I'm glad you resolved your problem.
